node/src/service/nonce_monitor.rs

```diff
@@ -0,0 +1,199 @@
+use std::time::Duration;
+
+use codec::Decode;
+use creditcoin_node_runtime::AccountId;
+use jsonrpc_core::{futures::channel::mpsc, Failure, Response, Success};
+use sc_client_api::Backend;
+use sc_service::{Arc, RpcHandlers, RpcSession};
+use sp_runtime::{app_crypto::Ss58Codec, offchain::OffchainStorage};
+use substrate_prometheus_endpoint::Registry;
+use thiserror::Error;
+
+use super::FullBackend;
+
+#[derive(Clone, Copy, PartialEq, Debug, Error)]
+enum ErrorKind {
+	#[error("Serde error")]
+	Serde,
+	#[error("RPC error")]
+	Rpc,
+	#[error("Codec error")]
+	Codec,
+}
+
+#[derive(Debug, Error)]
+#[error("{kind}: {value}")]
+struct Error {
+	kind: ErrorKind,
+	#[source]
+	value: Box<dyn std::error::Error + Send>,
+}
+
+macro_rules! error_fn {
+	($($fn_id: ident : $kind: ident),+ $(,)?) => {
+		impl Error {
+			$(
+				fn $fn_id(err: impl std::error::Error + Send + 'static) -> Self {
+					Self::new(ErrorKind::$kind, err)
+				}
+			)+
+		}
+	};
+}
+
+error_fn!(serde: Serde, rpc: Rpc, codec: Codec);
+
+impl Error {
+	fn new(kind: ErrorKind, err: impl std::error::Error + Send + 'static) -> Self {
+		Self { kind, value: Box::new(err) }
+	}
+}
+
+#[derive(Clone, PartialEq, Debug, Error)]
+#[error("{message}")]
+struct AdhocError {
+	message: String,
+}
+
+fn adhoc(message: impl Into<String>) -> AdhocError {
+	AdhocError { message: message.into() }
+}
```

Comment:
Do you like a more straight forward impl?

```suggestion
#[derive(Debug, Error)]
#[error("{self}")]
enum Error {
	Serde(sc_telemetry::serde_json::Error),
	JsonRpc(jsonrpc_core::Error),
	Rpc(String),
	Codec(codec::Error),
}
```

Reply:
Yeaah good point lol. I originally was thinking I might add extra info to the error like a "context" message (so just like an extra field where you can add a message with a description of what was happening when the error occurred) which led me to an error struct, but I never ended up adding extra info and then forgot to re-simplify the error type😅